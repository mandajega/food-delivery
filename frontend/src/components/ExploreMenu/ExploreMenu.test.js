import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import ExploreMenu from './ExploreMenu';
import { menu_list } from '../../assets/assets';

describe('ExploreMenu Component', () => {
  const mockSetCategory = jest.fn();
  const defaultProps = {
    category: 'All',
    setCategory: mockSetCategory,
  };

  const renderComponent = (props = defaultProps) => {
    const { asFragment } = render(<ExploreMenu {...props} />);
    return asFragment;
  };

  beforeEach(() => {
    mockSetCategory.mockClear();
  });

  test('renders the component with menu items', () => {
    const asFragment = renderComponent();
    expect(screen.getByText('Explore our menu')).toBeInTheDocument();
    expect(screen.getByText(/Choose from a diverse menu/i)).toBeInTheDocument();

    menu_list.forEach(item => {
      expect(screen.getByAltText(item.menu_name)).toBeInTheDocument();
      expect(screen.getByText(item.menu_name)).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  test('calls setCategory with correct arguments when a menu item is clicked', () => {
    renderComponent();
    const firstMenuItem = menu_list[0];
    fireEvent.click(screen.getByText(firstMenuItem.menu_name));

    expect(mockSetCategory).toHaveBeenCalledWith(expect.any(Function));

    const setCategoryMockImplementation = fn => {
      const newCategory = fn(defaultProps.category);
      expect(newCategory).toBe(firstMenuItem.menu_name);
    };

    expect(mockSetCategory.mock.calls[0][0]).toEqual(expect.any(Function));
    mockSetCategory.mock.calls[0][0](setCategoryMockImplementation);
  });

  test('applies active class to the selected category', () => {
    const props = {
      category: menu_list[0].menu_name,
      setCategory: mockSetCategory,
    };

    const asFragment = renderComponent(props);
    const activeImage = screen.getByAltText(menu_list[0].menu_name);
    expect(activeImage).toHaveClass('active');

    expect(asFragment()).toMatchSnapshot();
  });

  test('removes active class when clicking the same category (toggles to "All")', () => {
    const props = {
      category: menu_list[0].menu_name,
      setCategory: mockSetCategory,
    };

    const asFragment = renderComponent(props);
    fireEvent.click(screen.getByText(menu_list[0].menu_name));

    expect(mockSetCategory).toHaveBeenCalledWith(expect.any(Function));

    const setCategoryMockImplementation = fn => {
      const newCategory = fn(props.category);
      expect(newCategory).toBe('All');
    };

    mockSetCategory.mock.calls[0][0](setCategoryMockImplementation);

    expect(asFragment()).toMatchSnapshot();
  });
});
