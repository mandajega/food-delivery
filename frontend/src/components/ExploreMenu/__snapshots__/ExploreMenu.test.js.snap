// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExploreMenu Component applies active class to the selected category 1`] = `
<DocumentFragment>
  <div
    class="explore-menu"
    id="explore-menu"
  >
    <h1>
      Explore our menu
    </h1>
    <p
      class="explore-menu-text"
    >
      Choose from a diverse menu featuring a delectable array of dishes. Our mission is to satisfy your cravings and elevate your dining experience, one delicious meal at a time.
    </p>
    <div
      class="explore-menu-list"
    >
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Salad"
          class="active"
          src="test-file-stub"
        />
        <p>
          Salad
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Rolls"
          class=""
          src="test-file-stub"
        />
        <p>
          Rolls
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Deserts"
          class=""
          src="test-file-stub"
        />
        <p>
          Deserts
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Sandwich"
          class=""
          src="test-file-stub"
        />
        <p>
          Sandwich
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Cake"
          class=""
          src="test-file-stub"
        />
        <p>
          Cake
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Pure Veg"
          class=""
          src="test-file-stub"
        />
        <p>
          Pure Veg
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Pasta"
          class=""
          src="test-file-stub"
        />
        <p>
          Pasta
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Noodles"
          class=""
          src="test-file-stub"
        />
        <p>
          Noodles
        </p>
      </div>
    </div>
    <hr />
  </div>
</DocumentFragment>
`;

exports[`ExploreMenu Component removes active class when clicking the same category (toggles to "All") 1`] = `
<DocumentFragment>
  <div
    class="explore-menu"
    id="explore-menu"
  >
    <h1>
      Explore our menu
    </h1>
    <p
      class="explore-menu-text"
    >
      Choose from a diverse menu featuring a delectable array of dishes. Our mission is to satisfy your cravings and elevate your dining experience, one delicious meal at a time.
    </p>
    <div
      class="explore-menu-list"
    >
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Salad"
          class="active"
          src="test-file-stub"
        />
        <p>
          Salad
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Rolls"
          class=""
          src="test-file-stub"
        />
        <p>
          Rolls
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Deserts"
          class=""
          src="test-file-stub"
        />
        <p>
          Deserts
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Sandwich"
          class=""
          src="test-file-stub"
        />
        <p>
          Sandwich
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Cake"
          class=""
          src="test-file-stub"
        />
        <p>
          Cake
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Pure Veg"
          class=""
          src="test-file-stub"
        />
        <p>
          Pure Veg
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Pasta"
          class=""
          src="test-file-stub"
        />
        <p>
          Pasta
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Noodles"
          class=""
          src="test-file-stub"
        />
        <p>
          Noodles
        </p>
      </div>
    </div>
    <hr />
  </div>
</DocumentFragment>
`;

exports[`ExploreMenu Component renders the component with menu items 1`] = `
<DocumentFragment>
  <div
    class="explore-menu"
    id="explore-menu"
  >
    <h1>
      Explore our menu
    </h1>
    <p
      class="explore-menu-text"
    >
      Choose from a diverse menu featuring a delectable array of dishes. Our mission is to satisfy your cravings and elevate your dining experience, one delicious meal at a time.
    </p>
    <div
      class="explore-menu-list"
    >
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Salad"
          class=""
          src="test-file-stub"
        />
        <p>
          Salad
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Rolls"
          class=""
          src="test-file-stub"
        />
        <p>
          Rolls
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Deserts"
          class=""
          src="test-file-stub"
        />
        <p>
          Deserts
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Sandwich"
          class=""
          src="test-file-stub"
        />
        <p>
          Sandwich
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Cake"
          class=""
          src="test-file-stub"
        />
        <p>
          Cake
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Pure Veg"
          class=""
          src="test-file-stub"
        />
        <p>
          Pure Veg
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Pasta"
          class=""
          src="test-file-stub"
        />
        <p>
          Pasta
        </p>
      </div>
      <div
        class="explore-menu-list-item"
      >
        <img
          alt="Noodles"
          class=""
          src="test-file-stub"
        />
        <p>
          Noodles
        </p>
      </div>
    </div>
    <hr />
  </div>
</DocumentFragment>
`;
